src/relay/qnn/pass/qnn_lower.cc

```diff
@@ -0,0 +1,249 @@
+ */
+
+#include <tvm/relay/analysis.h>
+#include <tvm/relay/transform.h>
+#include <tvm/relay/op_attr_types.h>
+#include <tvm/relay/qnn/attrs.h>
+#include "../util.h"
+#include "../../pass/pattern_util.h"
+
+namespace tvm {
+namespace relay {
+namespace qnn {
+/*!
+ * \brief namespace of qnn lower pass.
+ *
+ * Use namespace to reduce potential naming conflict.
+ */
+namespace qnn_lower {
+
+using runtime::TypedPackedFunc;
+
+// Lowering of qnn.requantize op
+
+/*
+ * Converts a floating point number so that it can be represented by integers.
+ * The representation is
+ *      float_number = (significand) * 2^(exponent)
+ *
+ * The significand is a number between 0.5 and 1. This is represented
+ * by an integer number. For example, if it is int32, then the decimal point
+ * exists between bit 31 and 30 from LSB (or between first and second bit from
+ * the left).
+ *
+ * Some examples are
+ *           0.25 = (0.5) * 2^(-1)
+ *           0.125 = (0.5) * 2^(-2)
+ *
+ * Credit to TFLite reference implementation.
+ */
+std::pair<int, int> GetFixedPointMultiplierShift(double double_multiplier,
+    const DataType& idtype) {
+  int significand, exponent;
+  int idtype_bits = idtype.bits();
+
+  // Get the significand (significand) and exponent (exponent)
+  double significand_d = std::frexp(double_multiplier, &exponent);
+
+  // Convert the double significand to int significand.
+  significand_d = std::round(significand_d * (1ll << (idtype_bits - 1)));
+  auto significand_int64 = static_cast<int64_t>(significand_d);
+  CHECK_LE(significand_int64, (1ll << (idtype_bits - 1)));
+  if (significand_int64 == (1ll << (idtype_bits - 1))) {
+    significand_int64 /= 2;
+    ++exponent;
+  }
+  CHECK_LE(significand_int64, std::numeric_limits<int>::max());
+  significand = static_cast<int>(significand_int64);
+  return std::pair<int, int>(significand, exponent);
+}
+
+/*
+ * Requantization using only integer computation. Here, the computation is
+ * converted to a fixed point computation by computing output multiplier and
+ * shift. This is useful, if the target device does not support/have very
+ * expensive floating point computations.
+ *
+ * Original compuation is scale_fp32 * quantized_tensor.  To convert into
+ * integer computation, the multiplication with fp32 scalar can be replaced by
+ * multiplication with an int value and then right shifting the result. This
+ * approximates the floating point computation with a fixed point computation.
+ *
+ * The whole computation this can be broken down into following steps
+ * 1) Calculate the integer multiplier and integer shift.
+ * 2) Subtract the input integer point.
+ * 3) Multiply the integer fixed point multiplier with quantized tensor.
+ * 4) Round the result.
+ * 5) Right shift the result.
+ * 6) Add the output_zero_point.
+ * 7) Cast to the out_dtype.
+ *
+ */
+Expr RequantizeLower(const Expr& input_tensor,
+    const RequantizeAttrs* param, const DataType& idtype,
```

Comment:
idtype == input Tensor data type ?

Reply:
Shouldn't the inputs to this function get documented as per our coding standards  ?

Reply:
Just added, resolving this one.

Reply:
If *idtype The dtype of the input tensor*, we need this per TVM coding conversion?